# Recap 

In this part of the beginner series, we will build upon the code from <a href="https://www.fantasydatapros.com/cfb/blog/beginner/1">the last post</a>, introducing more layers and moving parts. We will discuss the Python namespace and how Python keeps track of all the objects we have defined in our program. Additionally, we will learn a little bit about libraries and how we can use them to import code into our namespace (although we won't be importing anything just yet) and built-in objects (which we've actually already used). We will also cover user-defined functions and functions in general.

# Built-in objects and the namespace

Built-in functions are functions (we touched briefly on this last time) that exist naturally in the Python <i>namespace</i>. The <a target="_blank" href="https://www.geeksforgeeks.org/namespaces-and-scope-in-python/">namespace</a> sounds very arcane, but essentially it is a collection of already defined names for things that we can use in our code. 

<h1>Namespaces</h1>

I asked chatGPT to explain the namespace, to which it replied:

<i>Python namespace is a collection of objects (variable, function, class, etc.) that can be accessed by a unique name. It's basically a way for Python to keep track of all the different objects that you have defined in your program, and make sure that each object has a unique name. This helps to prevent confusion and name collisions between different objects in your program.</i>

We can import things in to our namespace, as we'll see in later posts, to bring in more items (functions and classes) to write our code. Importing external code will be useful when we start using libraries like pandas, Python's data manipulation library. A library is a bunch of code already written for you so you don't have to reinvent the wheel. When you import those libraries into your code you are then able to use them once you get them into the namespace.

But there are certain objects that already exist without any necessary importing. These are called "built-in" objects. 

One such built-in function here is the `print` function, which we used in the last post. 

In [1]:
print('Hello World!')

Hello World!


We touched only briefly on functions last time, but essentially in Python, functions are blocks of code that are defined once and can be used repeatedly throughout a program. Functions allow for efficient and organized code by breaking down tasks into smaller, reusable chunks. They can accept arguments <i>(inputs)</i> and return values <i>(outputs)</i>. Functions can also be used to create modules, which are collections of related functions that can be imported into other Python scripts (as discussed, we'll be exploring some of these modules/libraries in later posts like pandas). Functions are a key part of Python programming and are used to make code more modular, maintainable, and readable.


<h1>User-defined Functions</h1>

User defined functions in Python are functions that are created and defined by the user, rather than being pre-defined in the language itself. Using an example from college football, a user may define a function that calculates the yards per carry for a player in a given week. This function would take in parameters such as the player's name, their position, and the number of rushing yards and carries they had and would return the average yards per carry. This user defined function could then be called and used multiple times throughout the program to calculate average yards per carry for different players.

To write a user-defined function in Python using a college football example, you could follow these steps:

1. Define the function by using the keyword `def` followed by the function name and any parameters that will be passed to the function. For example, a function that calculates a player's average yards per carry might be defined like this:

In [2]:
def avg_yards_per_carry(yards, carries):

IndentationError: expected an indented block (1143961586.py, line 1)

2. Inside the function, write the code that will perform the desired calculation. In the example above, this would be the code that divides the player's total yards by their number of carries:

In [3]:
avg = yards / carries

NameError: name 'yards' is not defined

3. Return the result of the calculation by using the `return` keyword. In the example above, this would be the player's average yards per carry:

In [4]:
return avg

SyntaxError: 'return' outside function (2172458130.py, line 1)

4. Call the function in your code by using the function name followed by any necessary arguments. For example, to calculate the average yards per carry for a player who has rushed for 100 yards on 20 carries, you would write:

In [5]:
avg_yards = avg_yards_per_carry(100, 20)

NameError: name 'avg_yards_per_carry' is not defined

5. Print the result of the function call to see the result of the calculation. In the example above, this would be the player's average yards per carry:

In [6]:
print(avg_yards)  # Output: 5.0

NameError: name 'avg_yards' is not defined

Overall, a user-defined function in Python using a football example might look like this:

In [21]:
def avg_yards_per_carry(yards, carries):
    avg = yards / carries
    return avg

avg_yards = avg_yards_per_carry(100, 20)
print(avg_yards) 

5.0


Now let's take all of this new information and use it to make our example from the last post even better. How can we modularize our code in a way that makes it better? The answer is that we can write a function that takes in a player's name, number of carries  and rushing yards and outputs a human-readable string that tells us that player's yards per carry for the season.

In [18]:
def calculate_yards_per_carry(player_name, carries, rushing_yards):
        yards_per_carry = rushing_yards/ carries
        yards_per_carry_rounded = round(yards_per_carry, 2)
        print(player_name, 'had a yards per carry of', yards_per_carry_rounded)
        
calculate_yards_per_carry('Bijan Robinson', carries=258, rushing_yards=1580)


Bijan Robinson had a yards per carry of 6.12


This is our simple Python function that calculates the yards per carry of a player. The function takes in three arguments:
<ul>
<li>`player_name`, which is a string representing the name of the player;</li> 
<li>`carries`, which is an integer representing the number of carries the player had; and </li>
<li>`rushing_yards`, which is an integer representing the number of yards a player had while running the ball.</li>
</ul>
The function first calculates the yards per carry by dividing the number of rushing yards by the number of carries. It then rounds the yards per carry to two decimal places and prints out a string that includes the player's name and their yards per carry.

In the example provided, the function is called with the player name "Bijan Robinson", 258 carries, and 1580 rushing yards. This means that Bijan Robinson had a yards per carry of 6.124031007751938 ypc (ypc is a common abbreviation for the unit 'yards per carry'), which gets rounded to 6.12 and gets printed out as part of the string.

<h1>Positional and Keyword Arguments</h1>

Here you can also see that we explicitly defined arguments when passing them in. "Bijan Robinson" was a positional argument and carries and rushing yards were keyword arguments.

In programming, a keyword argument is a type of argument that is passed to a function or method in which the name of the argument is specified in the function call. This allows for greater clarity and readability of the code, as the name of the argument clearly indicates its purpose.

As an example, consider a function that calculates the average catches per game for a given college football player. This function might have a keyword argument called "player_name" that specifies the name of the player for whom the average catches per game should be calculated. The function could then be called like this:

In [19]:
calculate_avg_catches_per_game(player_name="Jackson Smith-Njigba")

NameError: name 'calculate_avg_catches_per_game' is not defined

In this example, the "player_name" keyword argument is used to specify the player for whom the average catches per game should be calculated. The use of a keyword argument makes the code more readable and easier to understand, as it is clear that the "player_name" argument is used to specify the player in question. Our example above is no different.

Let's circle back to introducing this new idea of user-defined functions in to our code.

In [20]:
players = [
    {
    "name": "Trey Benson",
    "team": "THE Florida State Seminoles",
    "carries": 141,
    "rushing_yards": 965
    },
    {
    "name": "Bijan Robinson",
    "team": "Texas Longhorns",
    "carries": 258,
    "rushing_yards": 1580
    },
    {
    "name": "Jahmyr Gibbs",
    "team": "Alabama Crimson Tide",
    "carries": 136, 
    "rushing_yards": 850
    }
]


for player in players:
    calculate_yards_per_carry(player['name'], carries=player['carries'], rushing_yards=player['rushing_yards'])

Trey Benson had a yards per carry of 6.84
Bijan Robinson had a yards per carry of 6.12
Jahmyr Gibbs had a yards per carry of 6.25


As you can see here, we bring in our list of `dictionary` objects, each containing information about 3 college football RB's - Trey Benson, Bijan Robinson, and Jahmyr Gibbs.

We then iterate through each of our player objects and pass in their data to the function, which provides the desired output! Compare this to our previous post and you could probably notice that the code is a lot cleaner now.

<h1>Concluding Thoughts</h1>

We covered a decent amount of theory in this post. To recap, we: 
<ul>
<li>Expanded our knowledge of functions</li>
<li>Acknowledged that some of these functions are built-in to the namespace</li>
<li>Some of them can be imported via external libraries (which we will cover in the future), and</li> 
<li>Some of these can be written by us as user-defined functions</li>
</ul>

We also covered in-depth what functions are and some other details like positional and keyword arguments.

Using all of this, we took our code from last time and made it slightly more <a href="https://stackoverflow.com/questions/25011078/what-does-pythonic-mean" target="_blank">Pythonic</a> (emphasis on slightly because at the moment, our code isn't all that more Pythonic yet, but we'll touch on this topic more in future posts. "Pythonic" code is essentially Python code that is written with good style/convention).

No groundbreaking analysis yet, but we're one step closer to do some real analysis in Python! 

Thanks for reading - You guys are all awesome and happy coding!